tests/infer/test_gradient.py

```diff
@@ -35,6 +35,92 @@ def DiffTrace_ELBO(*args, **kwargs):
     return Trace_ELBO(*args, **kwargs).differentiable_loss
 
 
+@pytest.mark.parametrize(
+    "reparameterized,has_rsample",
+    [(True, None), (True, False), (True, True), (False, None)],
+    ids=["reparam", "reparam-False", "reparam-True", "nonreparam"],
+)
+@pytest.mark.parametrize(
+    "Elbo",
+    [
+        Trace_ELBO,
+        TraceEnum_ELBO,
+    ],
+)
+def test_particle_gradient(Elbo, reparameterized, has_rsample):
+    pyro.clear_param_store()
+    data = torch.tensor([-0.5, 2.0])
+    Normal = dist.Normal if reparameterized else fakes.NonreparameterizedNormal
+
+    def model():
+        with pyro.plate("data", len(data)) as ind:
+            x = data[ind]
+            z = pyro.sample("z", Normal(0, 1))
+            pyro.sample("x", Normal(z, 1), obs=x)
+
+    def guide():
+        scale = pyro.param("scale", lambda: torch.tensor([1.0]))
+        with pyro.plate("data", len(data)):
+            loc = pyro.param("loc", lambda: torch.zeros(len(data)), event_dim=0)
+            z_dist = Normal(loc, scale)
+            if has_rsample is not None:
+                z_dist.has_rsample_(has_rsample)
+            pyro.sample("z", z_dist)
+
+    elbo = Elbo(
+        max_plate_nesting=1,  # set this to ensure rng agrees across runs
+        num_particles=1,
```

Comment:
From this test alone it is unclear whether there is a bug in _stochastic_ gradient estimation; rather it is clear only that `Trace_ELBO` disagrees with a particular hand-coded estimator (at a single sample).  That is, it still seems possible that `Trace_ELBO` might agree with your hand-coded estimator in expectation, but disagree at each sample.  To confirm this is a true inference bug, we could also try setting `num_particles=100000, vectorize_particles=True` and see whether the _expected_ gradients coming out of `Trace_ELBO` and `TraceEnum_ELBO` agree.  WDYT?

Reply:
You are right! The missing term is just a score function and its expectation is zero (which is cleverly removed by `Trace_ELBO`). So the expected gradients of `Trace_ELBO` and `TraceEnum_ELBO` agree (as tested in `test_subsample_gradient`).

I fixed the hand-coded formula for `Trace_ELBO` so the test should pass now. @fritzo do you want to add the test to the repo? If not, I'll close the PR.